server-phoenix/lib/helios_web/controllers/web_hooks/slack_controller.ex

```diff
@@ -1,33 +1,175 @@
 defmodule HeliosWeb.WebHooks.SlackController do
   use HeliosWeb, :controller
-  alias Helios.{Repo, Event}
+  alias Helios.{Repo, Event, Location, Announcement}
 
   def handle(conn, params) do
-    if params["type"] == "url_verification" do
-      send_resp(conn, 200, params["challenge"])
-    else
-      event = params["event"]
-
-      unless Event
-             |> Event.slack_messages()
-             |> Event.with_external_id(event["event_ts"])
-             |> Repo.exists?() do
-        Repo.insert!(%Event{
-          source: :slack_message,
-          external_id: event["event_ts"]
-        })
-        |> publish
-      end
-
-      send_resp(conn, 200, "OK")
+    case params["type"] do
+      "url_verification" ->
+        send_resp(conn, 200, params["challenge"])
+
+      "event_callback" ->
+        event = params["event"]
+
+        unless allow_event?(event) do
+          send_resp(conn, 200, "Skipping disallowed event")
+        else
+          case slack_event(event) do
+            {:ok, _} ->
+              send_resp(conn, 200, "OK")
+
+            {:error, message} ->
+              send_slack_message(event["user"], message)
+              send_resp(conn, 200, "OK")
+          end
+        end
+
+      _ ->
+        send_resp(conn, 200, "Invalid request type")
     end
   end
 
-  defp publish(event) do
+  defp allow_event?(event) do
+    subtypes =
+      ~w(file_comment file_mention file_share me_message message_replied thread_broadcast)
+
+    (event["type"] == "message" || event["type"] == "app_mention") &&
+      (event["subtype"] == nil || Enum.member?(subtypes, event["subtype"]))
+  end
+
+  defp slack_event(event) do
+    case event["type"] do
+      "message" ->
+        unless Event
+               |> Event.slack_messages()
+               |> Event.with_external_id(event["event_ts"])
+               |> Repo.exists?() or message_has_mention?(event) do
+          Repo.insert!(%Event{
+            source: :slack_message,
+            external_id: event["event_ts"]
+          })
+          |> publish_event
+        else
+          {:ok, "Event Skipped"}
+        end
+
+      "app_mention" ->
+        case parse_announcement(event) do
```

Comment:
seems like a lot of stuff is happening here. Maybe we could split it up?

Reply:
Yeah, I can do that. You're talking about splitting up the commit, right?